### PR TITLE
Fix fused cross-entropy backward with broadcasted grad_output

### PR DESCRIPTION
**Pull request by sudhu2k**
# Description
The fused cross-entropy backward corrupts gradients when autograd passes a broadcasted `grad_output`. Reductions like `loss.sum()` produce a gradient that is logically all-ones but is physically a single element expanded with stride 0 (e.g. shape=(64, 2), stride=(0, 0), numel=128, storage holds 1 element).

cross_entropy_backward launches element_mul_kernel with a per-row stride of 1 if grad_output.numel() > 1 else 0. Since numel() > 1, it assumes a unit row stride and reads one value per row from contiguous memory. But the tensor only has one element in storage. Rows beyond the first read past the end of the allocation, yielding garbage that propagates as inf/NaN gradients 

The fix materializes a contiguous grad_output before the kernel launch so the per-row indexing is always valid.

Fixes: [Megatron Core_r0.17.0 Unit test failure](https://github.com/ROCm/Megatron-LM/actions/runs/27381139909/job/80947277363) 

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

In `transformer_engine/pytorch/triton/cross_entropy.py,` `cross_entropy_backward` now calls `grad_output = grad_output.contiguous()` (when numel() > 1) before launching `element_mul_kernel`, ensuring the assumed unit row stride matches the tensor's physical layout and preventing out-of-bounds reads for broadcasted/expanded gradients.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/triton/cross_entropy.py [hunk 1]
```diff
@@ -132,6 +132,11 @@ def cross_entropy_backward(
         n_rows = B * SQ
         BLOCK_SIZE = min(MAX_FUSED_SIZE, triton.next_power_of_2(V))
 
+        # element_mul_kernel indexes grad_output per row with unit stride, so a broadcasted
```
Comment (ipanfilo) [inline transformer_engine/pytorch/triton/cross_entropy.py:135]: It looks good. Do you have an idea if other scenarios might have similar problem except cross_entropy?

Comment (sudhu2k) [reply transformer_engine/pytorch/triton/cross_entropy.py:135]: The other triton kernels seems to be reading the stride.
https://github.com/ROCm/TransformerEngine/blob/e66f431af70df4bd73d996da51842ef512fb5051/transformer_engine/pytorch/triton/permutation.py#L341-L345

https://github.com/ROCm/TransformerEngine/blob/e66f431af70df4bd73d996da51842ef512fb5051/transformer_engine/pytorch/triton/pad.py#L25-L41
cross_entropy was affected since there's a high chance of the losses to be broadcasted, but I don't think other layers' kernels would be affected by this issue. 

I noticed that rmsnorm kernels called 
```
    dz_ = dz.contiguous()
    x_ = x.contiguous()
    rsigma_ = rsigma.contiguous()
    gamma_ = gamma.contiguous()
```
but layernorm triton kernels didn't, so I've added these lines to the layernorm triton kernel wrapper.

cast.py — input_tensor = tensor.contiguous(). Safe.
cast_transpose.py — reads dynamic .stride(0)/.stride(1), uses .contiguous(), and even asserts stride(0)==1 or stride(1)==1. Safe.

**Review by ipanfilo — APPROVED** [review]
Pending CI pass
